saleor/saleor
# Add brand field to app manifest

I want to merge this change because it adds brand information to the app manifest and app model. For now, it consists only of the default image logo, but it could be easily extended with additional information in the future. I would like to use this PR as a base to discuss the following:

1. New optional field in the App manifest: `brand`
    ```
    brand: {
      logo: {
        default: string // url
      },
    }
    ```

2. Icon image validation steps:
    1. Fetch the image from `brand.logo.default` URL
    2. Check and accept only PNG image format as input
    3. Check if the image has the minimum required size (256px) and is square

3. Thumbnails generation of icon images is possible only to formats `WebP` and `PNG` (lossless compression)
`
4. Mutation `appFetchManifest` fetches the logo image, validates it, and returns its thumbnail on the fly as a string in the form of base64 encoded data URL (for example: `data:image/webp;base64,UklGRtoPAABXRUJQV...`). A new optional field `brand` of type `AppManifestBrand` is added to the `Manifest` object type:
```graphql
enum IconThumbnailFormatEnum {
  ORIGINAL
  WEBP
}

type AppManifestBrand @doc(category: "Apps") {
  logo: AppManifestBrandLogo!
}

type AppManifestBrandLogo @doc(category: "Apps") {
  default(size: Int format: IconThumbnailFormatEnum = ORIGINAL): String!
}
```
Brand data can be used by the dashboard for displaying Manifest preview: 
<img width="1444" alt="manifest_preview" src="https://user-images.githubusercontent.com/6453010/227476837-2e4aefa2-485a-43dc-a102-a7e6a85ecfed.png">



6. Mutation `appInstall` fetches the logo image, validates it, and stores it in the Saleor's file storage. `ImageField` is added to `App` and `AppInstallation` models, and `App` and `AppInstallation` graphQL object types have a new optional field `brand` of type `AppBrand`. Thanks to that dashboard can display apps and app installations icons with logos. 
```graphql
"""Represents the app's brand data."""
type AppBrand @doc(category: "Apps") {
  logo: AppBrandLogo!
}

type AppBrandLogo @doc(category: "Apps") {
  """URL to the default logo image."""
  default(size: Int format: IconThumbnailFormatEnum = ORIGINAL): String!
}
```

### TODO
* [x] Add `uuid` field to `App` and `AppInstallation` models
* [x] Handle thumbnail generation for `App` and `AppInstallation`
* [x] Move and reformat code to avoid duplication (image validation, etc.)



# Impact

* [x] Documentation needs to be updated

# Pull Request Checklist



* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database migration files are up to date
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog
